Source/WebCore/page/EventHandler.cpp

```diff
@@ -4337,49 +4337,105 @@ void EventHandler::defaultBackspaceEventHandler(KeyboardEvent& event)
+    auto* platformEvent = event.underlyingPlatformEvent();
+
+    if (animator && platformEvent)
+        return animator->beginKeyboardScrollGesture(*platformEvent, m_frame.page());
+
+    return false;
+}
+
+bool EventHandler::startKeyboardScrollingOnFrame(KeyboardEvent& event)
+{
+    RefPtr frame = &m_frame;
+    RefPtr view = frame->view();
+    if (!view)
+        return false;
 
-    auto* scrollableArea = enclosingScrollableArea(node);
+    auto* animator = view->scrollAnimator().keyboardScrollingAnimator();
+    return beginKeyboardScrollGesture(animator, event);
+}
+
+bool EventHandler::keyboardScrollOnRenderBoxLayer(KeyboardEvent& event, RenderBox* renderBox)
+{
+    auto* scrollableArea = renderBox->layer() ? renderBox->layer()->scrollableArea() : nullptr;
     if (!scrollableArea)
-        return nullptr;
+        return false;
 
-    return scrollableArea->scrollAnimator().keyboardScrollingAnimator();
+    auto* animator = scrollableArea->scrollAnimator().keyboardScrollingAnimator();
+    return beginKeyboardScrollGesture(animator, event);
 }
 
-void EventHandler::stopKeyboardScrolling()
+bool EventHandler::keyboardScrollOnRenderBox(KeyboardEvent& event, RenderBox* renderBox)
 {
-    auto* animator = keyboardScrollingAnimatorForFocusedNode();
-    if (animator)
-        animator->handleKeyUpEvent();
+    if (keyboardScrollOnRenderBoxLayer(event, renderBox))
+        return true;
+
+    RenderBlock* nextRenderBlock = renderBox->containingBlock();
+
+    if (nextRenderBlock && !nextRenderBlock->isRenderView())
+        return keyboardScrollOnRenderBox(event, nextRenderBlock);
+
+    return false;
 }
 
-bool EventHandler::startKeyboardScrolling(KeyboardEvent& event)
+bool EventHandler::keyboardScrollOverflow(KeyboardEvent& event, Node* startingNode)
+{
+    RefPtr node = startingNode;
+
+    if (!node)
+        node = m_frame.document()->focusedElement();
+
+    if (!node)
+        node = m_mousePressNode.get();
+
+    if (node) {
+        auto renderer = node->renderer();
+        RenderBox& renderBox = renderer->enclosingBox();
+        if (renderer && !renderer->isListBox() && keyboardScrollOnRenderBox(event, &renderBox))
+            return true;
+    }
+    return false;
+}
+
+bool EventHandler::keyboardScrollRecursively(KeyboardEvent& event, Node* startingNode)
 {
     Ref protectedFrame = m_frame;
 
+    m_frame.document()->updateLayoutIgnorePendingStylesheets();
+
     if (!m_frame.settings().eventHandlerDrivenSmoothKeyboardScrollingEnabled())
         return false;
 
-    auto* animator = keyboardScrollingAnimatorForFocusedNode();
-    auto* platformEvent = event.underlyingPlatformEvent();
+    if (keyboardScrollOverflow(event, startingNode))
+        return true;
 
-    if (animator && platformEvent)
-        return animator->beginKeyboardScrollGesture(*platformEvent);
+    if (startKeyboardScrollingOnFrame(event))
+        return true;
 
-    return false;
+    RefPtr frame = &m_frame;
+    frame = frame->tree().parent();
```

Comment:
Why doesn't this use `protectedFrame`?

Reply:
protectedFrame already exists at the top of the method, and no one else in EventHandler uses it directly, are we changing that?